File: urbanairship-core/src/main/java/com/urbanairship/actions/tags/BaseTagsAction.java
/* Copyright Airship and Contributors */

package com.urbanairship.actions.tags;

import androidx.annotation.NonNull;

import com.urbanairship.UAirship;
import com.urbanairship.actions.Action;
import com.urbanairship.actions.ActionArguments;
import com.urbanairship.actions.ActionResult;
import com.urbanairship.json.JsonValue;
import com.urbanairship.push.PushManager;
import com.urbanairship.util.UAStringUtil;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * Abstract tag action class.
 */
abstract class BaseTagsAction extends Action {

    /**
     * JSON key for channel tag group changes.
     */
    private static final String CHANNEL_KEY = "channel";

    /**
     * JSON key for named user tag group changes.
     */
    private static final String NAMED_USER_KEY = "named_user";

    /**
     * JSON key for device tags.
     */
    private static final String DEVICE_KEY = "device";

    /**
     * Gets the push manager
     *
     * @return A push manager instance
     */
    @NonNull
    protected PushManager getPushManager() {
        return UAirship.shared().getPushManager();
    }

    @Override
    public boolean acceptsArguments(@NonNull ActionArguments arguments) {
        if (arguments.getValue().isNull()) {
            return false;
        }

        if (arguments.getValue().getString() != null) {
            return true;
        }

        if (arguments.getValue().getList() != null) {
            return true;
        }

        return arguments.getValue().getMap() != null;

    }

    @NonNull
    @Override
    public ActionResult perform(@NonNull ActionArguments arguments) {

        if (arguments.getValue().getString() != null) {
            Set<String> tags = new HashSet<>();
            tags.add(String.valueOf(arguments.getValue().getString()));
            applyChannelTags(tags);
        }

        if (arguments.getValue().getList() != null) {
            Set<String> tags = new HashSet<>();
            for (JsonValue tag : arguments.getValue().getList()) {
                if (tag.getString() != null) {
                    tags.add(tag.getString());
                }
            }

            applyChannelTags(tags);
        }

        if (arguments.getValue().getMap() != null) {

            // Channel Tag Groups
            Map<String, Set<String>> tagsMap = new HashMap<>();
            for (Map.Entry<String, JsonValue> entry : arguments.getValue().getMap().opt(CHANNEL_KEY).optMap().getMap().entrySet()) {
                String group = entry.getKey();
                Set<String> tags = new HashSet<>();
                for (JsonValue jsonValue : entry.getValue().optList().getList()) {
                    tags.add(jsonValue.getString());
                }

                if (!UAStringUtil.isEmpty(group) && !tags.isEmpty()) {
                    tagsMap.put(group, tags);
                }
            }

            if (!tagsMap.isEmpty()) {
                applyChannelTagGroups(tagsMap);
            }

            // Named User Tag Groups
            tagsMap = new HashMap<>();
            for (Map.Entry<String, JsonValue> entry : arguments.getValue().getMap().opt(NAMED_USER_KEY).optMap().getMap().entrySet()) {
                String group = entry.getKey();
                Set<String> tags = new HashSet<>();
                for (JsonValue jsonValue : entry.getValue().optList().getList()) {
                    tags.add(jsonValue.getString());
                }

                if (!UAStringUtil.isEmpty(group) && !tags.isEmpty()) {
                    tagsMap.put(group, tags);
                }
            }

            if (!tagsMap.isEmpty()) {
                applyNamedUserTagGroups(tagsMap);
            }

            // Device Tags
            Set<String> tags = new HashSet<>();
            for (JsonValue tag : arguments.getValue().getMap().opt(DEVICE_KEY).optList()) {
                if (tag.getString() != null) {
                    tags.add(tag.getString());
                }
            }

            if (!tags.isEmpty()) {
                applyChannelTags(tags);
            }
        }

        return ActionResult.newEmptyResult();
    }

    /**
     * Applies tag updates.
     *
     * @param tags The set of tags.
     */
    abstract void applyChannelTags(@NonNull Set<String> tags);

    /**
     * Applies channel tag group updates.
     *
     * @param tags The map of tag groups.
     */
    abstract void applyChannelTagGroups(@NonNull Map<String, Set<String>> tags);

    /**
     * Applies named user tag group updates.
     *
     * @param tags The map of tag groups.
     */
    abstract void applyNamedUserTagGroups(@NonNull Map<String, Set<String>> tags);

}

File: urbanairship-core/src/main/java/com/urbanairship/actions/tags/BaseTagsAction.kt
/* Copyright Airship and Contributors */
package com.urbanairship.actions.tags

import com.urbanairship.actions.Action
import com.urbanairship.actions.ActionArguments
import com.urbanairship.actions.ActionResult
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow

/**
 * Abstract tag action class.
 */
public abstract class BaseTagsAction : Action() {

    private val mutableMutations: MutableSharedFlow<TagActionMutation> =  MutableSharedFlow()

    /**
     * Tag mutations
     */
    public val mutationsFlow: SharedFlow<TagActionMutation> = mutableMutations.asSharedFlow()

    override fun acceptsArguments(arguments: ActionArguments): Boolean {
        if (arguments.value.isNull) {
            return false
        }

        if (arguments.value.string != null) {
            return true
        }

        if (arguments.value.list != null) {
            return true
        }

        return arguments.value.map != null
    }

    override fun perform(arguments: ActionArguments): ActionResult {
        if (arguments.value.string != null) {
            arguments.value.string?.let {
                val tags = setOf(it)
                applyChannelTags(tags)
                mutableMutations.tryEmit(TagActionMutation.ChannelTags(tags))
            }
        }

        if (arguments.value.list != null) {
            val tags = arguments.value.list?.mapNotNull { it.string }?.toSet()
            if (!tags.isNullOrEmpty()) {
                applyChannelTags(tags)
                mutableMutations.tryEmit(TagActionMutation.ChannelTags(tags))
            }
        }

        if (arguments.value.map != null) {
            // Channel Tag Groups
            arguments.value.map?.get(CHANNEL_KEY)?.optMap()?.let { map ->
                val tagGroups = mutableMapOf<String, Set<String>>()
                map.forEach {
                    val tags = it.value.list?.mapNotNull { it.string }
                    if (!tags.isNullOrEmpty()) {
                        tagGroups[it.key] = tags.toSet()
                    }
                }
                if (tagGroups.isNotEmpty()) {
                    applyChannelTagGroups(tagGroups)
                    mutableMutations.tryEmit(TagActionMutation.ChannelTagGroups(tagGroups))
                }
            }

            // Contact Tag Groups
            arguments.value.map?.get(NAMED_USER_KEY)?.optMap()?.let { map ->
                val tagGroups = mutableMapOf<String, Set<String>>()
                map.forEach {
                    val tags = it.value.list?.mapNotNull { it.string }
                    if (!tags.isNullOrEmpty()) {
                        tagGroups[it.key] = tags.toSet()
                    }
                }
                if (tagGroups.isNotEmpty()) {
                    applyContactTagGroups(tagGroups)
                    mutableMutations.tryEmit(TagActionMutation.ContactTagGroups(tagGroups))
                }
            }

            // Device Tags
            arguments.value.map?.get(DEVICE_KEY)?.optList()?.let { list ->
                val tags = list.mapNotNull { tag -> tag.string }.toSet()
                if (tags.isNotEmpty()) {
                    applyChannelTags(tags)
                    mutableMutations.tryEmit(TagActionMutation.ChannelTags(tags))
                }
            }
        }

        return ActionResult.newEmptyResult()
    }

    /**
     * Applies tag updates.
     *
     * @param tags The set of tags.
     */
    internal abstract fun applyChannelTags(tags: Set<String>)

    /**
     * Applies channel tag group updates.
     *
     * @param tags The map of tag groups.
     */
    internal abstract fun applyChannelTagGroups(tags: Map<String, Set<String>>)

    /**
     * Applies contact tag group updates.
     *
     * @param tags The map of tag groups.
     */
    internal abstract fun applyContactTagGroups(tags: Map<String, Set<String>>)

    internal companion object {

        /**
         * JSON key for channel tag group changes.
         */
        private const val CHANNEL_KEY = "channel"

        /**
         * JSON key for named user tag group changes.
         */
        private const val NAMED_USER_KEY = "named_user"

        /**
         * JSON key for device tags.
         */
        private const val DEVICE_KEY = "device"
    }
}
